packages/expression-evaluator/lib/evaluators/ContextualizedEvaluator.ts

```diff
@@ -0,0 +1,147 @@
+import type { FunctionBusType, IActionContext, TermFunctionBusType } from '@comunica/types';
+import type * as RDF from '@rdfjs/types';
+import { LRUCache } from 'lru-cache';
+import type { Algebra as Alg } from 'sparqlalgebrajs';
+import * as E from '../expressions';
+import { expressionToVar } from '../functions/Helpers';
+import type { FunctionArgumentsCache } from '../functions/OverloadTree';
+import { AlgebraTransformer } from '../transformers/AlgebraTransformer';
+import type { ITimeZoneRepresentation } from '../util/DateTimeHelpers';
+import { extractTimeZone } from '../util/DateTimeHelpers';
+import * as Err from '../util/Errors';
+import type { SuperTypeCallback, TypeCache, ISuperTypeProvider } from '../util/TypeHandling';
+
+export type AsyncExtensionFunction = (args: RDF.Term[]) => Promise<RDF.Term>;
+export type AsyncExtensionFunctionCreator = (functionNamedNode: RDF.NamedNode) => AsyncExtensionFunction | undefined;
+
+export interface IAsyncEvaluatorContext {
+  extensionFunctionCreator?: AsyncExtensionFunctionCreator;
+  now?: Date;
+  baseIRI?: string;
+  typeCache?: TypeCache;
+  getSuperType?: SuperTypeCallback;
+  functionArgumentsCache?: FunctionArgumentsCache;
+  defaultTimeZone?: ITimeZoneRepresentation;
+  actionContext: IActionContext;
+  mediatorQueryOperation: MediatorQueryOperation;
+  mediatorTermFunction: TermFunctionBusType;
+  mediatorFunction: FunctionBusType;
+}
+
+/**
+ * This class provides evaluation functionality to already transformed expressions.
+ * It also holds all context items needed for evaluating functions.
+ */
+export class ContextualizedEvaluator {
+  protected readonly transformer: AlgebraTransformer;
+
+  private readonly subEvaluators: Record<string,
+  (expr: E.Expression, mapping: RDF.Bindings) => Promise<E.Term> | E.Term> =
+      {
+        [E.ExpressionType.Term]: this.term.bind(this),
+        [E.ExpressionType.Variable]: this.variable.bind(this),
+        [E.ExpressionType.Operator]: this.evalFunction.bind(this),
+        [E.ExpressionType.SpecialOperator]: this.evalFunction.bind(this),
+        [E.ExpressionType.Named]: this.evalFunction.bind(this),
+        [E.ExpressionType.Existence]: this.evalExistence.bind(this),
+        [E.ExpressionType.Aggregate]: this.evalAggregate.bind(this),
+        [E.ExpressionType.AsyncExtension]: this.evalFunction.bind(this),
+      };
+
+  // Context items
+  public readonly extensionFunctionCreator?: AsyncExtensionFunctionCreator;
+  public readonly now: Date;
+  public readonly baseIRI?: string;
+  public readonly functionArgumentsCache: FunctionArgumentsCache;
+  public readonly superTypeProvider: ISuperTypeProvider;
+  public readonly defaultTimeZone: ITimeZoneRepresentation;
+  public readonly actionContext: IActionContext;
+  public readonly mediatorQueryOperation: MediatorQueryOperation;
+  public readonly mediatorTermFunction: TermFunctionBusType;
+  public readonly mediatorFunction: FunctionBusType;
```

Comment:
@rubensworks I already mentioned for functions like `=` we need the sync version. (I call this a `termFunction` because the function executes over evaluated terms). I suggest having 2 busses like this. Has a subgroup of the actors of the other one. (This means that all functionActors are registered on the `functionBus` and some of them are also registered on the `termFunctionBus`.) Do you think this is okey, or do you have a better way in mind?)

Reply:
> I already mentioned for functions like = we need the sync version. (I call this a termFunction because the function executes over evaluated terms)

Doesn't both async and sync both evaluate over terms?

> I suggest having 2 busses like this. Has a subgroup of the actors of the other one. (This means that all functionActors are registered on the functionBus and some of them are also registered on the termFunctionBus.) Do you think this is okey, or do you have a better way in mind?)

Sounds a bit confusing.
Would it be doable to have one bus, but let the IAction interface define `sync: boolean` as a parameter?
This would allow some actors (the async ones) to reject actions that require sync evaluation.

This would also solve the naming issue above. All of them would just be functions then.

Reply:
So the difference is whether the arguments still need to be evaluated. Each function has an `apply` function that gets `args: E.Expression[]`. `E.Expression[]` can contain operations that would need an async evaluation. Some functions have a definition on terms (these used to be called `regularFunctions`, and are the ones defined using an overloadTree). Those functions have the same `apply` function and an additional `applyOnTerms` function that takes `args: E.TermExpression[]`. 

That being said, I would not refer to them as being sync or async. But rather also having the option of passing terms directly.

(Note: this requires a shift away from thinking functions get executed on terms. They mostly do, but not all, some need control over the evaluation of their arguments (used to call `specialFunctions`). The ones that don't need control over the evaluation are just a special case of function. The control flow is just way prettier defining functions this way).

Reply:
The name of the flag could be something like `evaluatesOnTerm`? Or something else?

Reply:
I'm not really following to be honest. Can you link me the code towards these two different function interfaces?

Reply:
https://github.com/comunica/comunica/pull/1283/files#diff-99c32e329f466d175ad0920b234c18090aeb77c4a82bd7cda67f80070a9f4727R14 and https://github.com/comunica/comunica/pull/1283/files#diff-99c32e329f466d175ad0920b234c18090aeb77c4a82bd7cda67f80070a9f4727R51

Reply:
Can you clarify very precisely why this difference exists? (I can't derive it from the jsdoc)
I see both have `exprEval` and `exprEval`, but only one has a `mapping`.
So is the main difference that one has a mappings context while the other has not?
Should this difference really exist? I.e., would it hurt if the other would _also_ have a mappings context, but just not use it?
If this is really needed, then this mappings context could be an inspiration for a name?

Reply:
The difference is some functions (e.g. `or`) need explicit control over the evaluation of their arguments. This means that they for example ignore exceptions thrown, or see exceptions as "false", or they need a lazy evaluation approach, ....

Other functions do not need this control. The functions that do not need that control can (mostly) be sync. What I suggest is that a function that doesn't care about the evaluation is just a special case of a function that does care. So, in principle both need the mappings, but one is just defined in a way where it doesn't care for argument evaluation and thus in its definition doesn't need the mapping. (And since we hold on to that definition, we can refer to it)

Reply:
> The functions that do not need that control can (mostly) be sync. 

So not all of them are sync?
Or do they?
If they are, the sync/async distinction still makes sense to me :-)

In any case, I would keep them in the same bus, and use some distinguishing flag in case we require only one type.
This flag name is something we can still discuss :-)